### PR TITLE
Async loading of bundles work with configurable string fetcher.

### PR DESCRIPTION
**Pull request by djg**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -1,4 +1,36 @@
...
+/// trait to provide loading of fluent resources, returning a `String`
+/// representating the resource.
+///
+/// Possible alternate names: FileSource, ResourceSource, StringLoader.
+///
+/// TODO: This should live in a fluent crate. But circular dependencies.
+#[async_trait]
+pub trait StringFetcher {
+    /// Return the `String` representation for `resource_id`. This version is
+    /// blocking. See `fetch_async`.
```
Comment (zbraniecki) [inline src/lib.rs]: nit: `fetch_async` -> `fetch`

File: src/lib.rs [hunk 2]
```diff
@@ -1,4 +1,36 @@
...
+    ///
+    /// Note: Possible alternate name: `read_sync`, `load_async`.  `fetch` works
+    /// is a generic way, meaning `get me the string for, I don't care how`.
```
Comment (zbraniecki) [inline src/lib.rs]: I like `fetch` and `fetch_sync` for the function that doesn't do logic, just fetches. I use `load` for cases which may cache, or perform some cache invalidation.

File: src/registry/asynchronous.rs [hunk 3]
```diff
@@ -1,46 +1,140 @@
...
+        GenerateBundles::new(self.clone(), lang_ids, resource_ids)
+    }
+
+    pub fn generate_bundles(
+        &self,
+        lang_ids: Vec<LanguageIdentifier>,
+        resource_ids: Vec<String>,
+    ) -> GenerateBundles {
+        let lang_ids = lang_ids.into_iter().collect();
+        let resource_ids = resource_ids.into_iter().collect();
```
Comment (zbraniecki) [inline src/registry/asynchronous.rs]: why do you do this here? Do you need to clone it?

File: src/registry/asynchronous.rs [hunk 4]
```diff
@@ -1,46 +1,140 @@
...
+}
+
+impl Stream for GenerateBundles {
+    type Item = FluentBundle;
+
+    fn poll_next(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Self::Item>> {
+        let this = &mut *self;
```
Comment (zbraniecki) [inline src/registry/asynchronous.rs:85]: can you document this? This is very non-trivial :)

File: src/registry/mod.rs [hunk 5]
```diff
@@ -1,61 +1,210 @@
...
+    //     langid: &'l LanguageIdentifier,
+    //     res_ids: impl IntoIterator<Item = &'l str>,
+    // ) -> impl Iterator<Item = Vec<&FileSource>>
+    // {
+    //     let resource_ids = res_ids.into_iter().collect::<Vec<_>>();
+    //     resource_ids
+    //         .into_iter()
+    //         .map(|res_id| self.generate_sources_for_file(langid, res_id))
+    //         .multi_cartesian_product()
+    // }
```
Comment (zbraniecki) [inline src/registry/mod.rs]: leftover?

File: src/registry/mod.rs [hunk 6]
```diff
@@ -1,61 +1,210 @@
...
+
+    pub fn get_source(&self, name: &str) -> Option<&FileSource> {
+        self.lock.iter().find(|&source| source.name == name)
+    }
+
+    pub fn generate_sources_for_file<'l>(
+        &'l self,
+        langid: &'l LanguageIdentifier,
+        res_id: &'l str,
+    ) -> impl Iterator<Item = &FileSource> {
```
Comment (zbraniecki) [inline src/registry/mod.rs:71]: do you need the type annotations if all uses the same?

File: src/source.rs [hunk 7]
```diff
@@ -1,62 +1,70 @@
...
+        match this {
+            Missing => None.into(),
+            Loaded(res) => Some(res.clone()).into(),
+            Loading(res) => Pin::new(res).poll(cx),
+        }
+    }
 }
 
+/// This is an implementation of a Gecko File Source.  It can be used to return
+/// Fluent resources from Omnija, etc.
```
Comment (zbraniecki) [inline src/source.rs]: I think this comment is misleading. This is a `FileSource` and can be anything. In Gecko we'll provide a `StringFetcher` and a set of `FileSource`'s, but this crate is more generic.

File: src/source.rs [hunk 8]
```diff
@@ -103,65 +115,45 @@ impl FileSource {
...
+    pub fn fetch_file_sync(&self, langid: &LanguageIdentifier, path: &str) -> RcResourceOption {
+        use ResourceStatus::*;
+
         let full_path = self.get_path(langid, &path);
 
-        let mut cache = self.cache.try_borrow_mut().unwrap();
-        let res = cache.entry(full_path.clone()).or_insert_with(|| {
-            (self.fetch_sync)(&full_path)
-                .expect("I/O Error")
+        // println!("[l10nregistry] fetch_file_sync: {:?}", full_path);
```
Comment (zbraniecki) [inline src/source.rs]: leftover? or do you want to leave it as `debug`?

File: src/source.rs [hunk 9]
```diff
@@ -103,65 +115,45 @@ impl FileSource {
...
-            ResourceStatus::Value(res) => Some(res.clone()),
-            ResourceStatus::Async(..) => {
-                println!("[l10nregistry] Attempting to synchronously load file {} while it's being loaded asynchronously.", &full_path);
-                cache.remove(&full_path);
-                drop(cache);
-                self.fetch_file_sync(langid, path)
+            Missing => None,
+            Loaded(res) => Some(res),
+            Loading(..) => {
+                panic!("[l10nregistry] Attempting to synchronously load file {} while it's being loaded asynchronously.", &full_path);
```
Comment (zbraniecki) [inline src/source.rs]: could we here `debug_print` and drop the async load, and replace it with a sync load?

File: src/source.rs [hunk 10]
```diff
@@ -103,65 +115,45 @@ impl FileSource {
...
-                        // My initial attempt to use it didn't work, but that may be just
-                        // my lack of experience with the API.
-                        .inspect(|res| set_resolved(cache, cloned_full_path, res.clone()))
-                        .boxed_local()
-                        .shared(),
-                )
-            })
-            .clone();
-        drop(cache);
+        // println!("[l10nregistry] fetch_file: {:?}", full_path);
```
Comment (zbraniecki) [inline src/source.rs]: debug log or drop this line?

File: src/source.rs [hunk 11]
```diff
@@ -186,7 +173,119 @@ impl std::fmt::Debug for FileSource {
...
+        let mut lock = self.entries.borrow_mut();
+        let entry = lock.get_mut(&path);
+        match entry {
+            Some(entry) => *entry = resource.clone().into(),
+            _ => panic!("Expected "),
+        }
+        resource
+    }
+
+    pub fn has_file(&self, full_path: &str) -> Option<bool> {
```
Comment (zbraniecki) [inline src/source.rs:229]: If I'm not mistaken the `None` here means `we don't know if the file exists - it's either currently being loaded, or we never tried`, can you add a doc string about it?

File: tests/registry.rs [hunk 12]
```diff
@@ -1,111 +1,101 @@
...
-    assert_eq!(i.next(), Some(vec![toolkit, toolkit]));
-    assert_eq!(i.next(), Some(vec![toolkit, browser]));
-    assert_eq!(i.next(), Some(vec![browser, toolkit]));
-    assert_eq!(i.next(), Some(vec![browser, browser]));
-    assert_eq!(i.next(), None);
+        // assert_eq!(i.next(), Some(vec![toolkit, toolkit]));
+        // assert_eq!(i.next(), Some(vec![toolkit, browser]));
+        // assert_eq!(i.next(), Some(vec![browser, toolkit]));
+        // assert_eq!(i.next(), Some(vec![browser, browser]));
+        // assert_eq!(i.next(), None);
```
Comment (zbraniecki) [inline tests/registry.rs]: leftover?

File: src/lib.rs [hunk 13]
```diff
@@ -1,4 +1,36 @@
...
 pub mod source;
+pub mod tokio;
+
+mod futures;
+
+/// Naming things is hard. The consumers of `fluent-fallback` implement this
+/// trait to provide loading of fluent resources, returning a `String`
+/// representating the resource.
+///
+/// Possible alternate names: FileSource, ResourceSource, StringLoader.
```
Comment (zbraniecki) [inline src/lib.rs]: How about `L10nFileFetcher`?

File: src/registry/asynchronous.rs (hunk 3 above)
Comment (djg) [reply src/registry/asynchronous.rs]: Left over code from a previous changes. Removed.

File: src/registry/mod.rs (hunk 5 above)
Comment (djg) [reply src/registry/mod.rs]: Yup

File: src/source.rs (hunk 9 above)
Comment (djg) [reply src/source.rs]: Or we could perform a sync load and not update the cache? I wasn't sure which was the best approach.

I wanted to be able to replace the loading state with loaded and fire the value to the waiting futures, but I'm not sure how to do that. A potential solution would be to store a `oneshot::Sender` and `Shared<oneshot::Receiver>`. The async loading future could then check that the state is still loading, and if so, `send` the result. The sync loading could do the same.

It's a little more complicated and I just wanted to get the system working. Could be leave this and open a issue to implement a more desirable behavior?

File: src/registry/asynchronous.rs (hunk 3 above)
Comment (djg) [reply src/registry/asynchronous.rs:47]: That code isn't needed anymore. Removed.

File: src/source.rs (hunk 8 above)
Comment (djg) [reply src/source.rs]: I've converted to `trace!` macros.

File: src/source.rs (hunk 10 above)
Comment (djg) [reply src/source.rs]: Converted to `trace!` macro

File: tests/registry.rs (hunk 12 above)
Comment (djg) [reply tests/registry.rs]: I removed the `source_permutations` API since it wasn't used except for testing.

File: src/registry/mod.rs (hunk 6 above)
Comment (djg) [reply src/registry/mod.rs:71]: With out the `'l` I get compilation errors:

```
lifetime mismatch
...but data from `langid` is returned hererustc(E0623)
mod.rs(69, 17): this parameter and the return type are declared with different lifetimes...
mod.rs(71, 10): 
mod.rs(71, 10): Error originated from macro here
```

File: src/source.rs (hunk 11 above)
Comment (djg) [reply src/source.rs:229]: Done

**Pull request merged**